### ISSUE TITLE
Grid system

### ISSUE THREAD

**Issue by drolsen** [connectivedx/fuzzy-chainsaw#110]
Knowing that we are not trying to make FC a treasure-chest of goodies but rather a starting point that we add features to as needed; I think a baseline grid system is still very much needed. Without one we can expect (regardless how hard we try) responsive approaches to become fractured from project to project, and things will get messy fast.

The old Phoenix grid system was nice and has been good to us, but I think it might be time to upgrade. I've been looking around as some postcss ones, and the one I'm most excited to use (as well as most of the postcss community) is the Lost grid system.

# **Lost**
https://www.npmjs.com/package/lost
https://www.npmjs.com/package/lost#basic-columns
https://www.youtube.com/user/LevelUpTuts/search?query=lost

The largest difference between Phoenix grid system and Lost  is instead of expressing column sizes by whole numbers, we express them in fractions. This means we are no longer locked in globally to a fixed number of columns. This gives us lots more room to be flexible in grid alignment for designers and never be on the hook to explain why we can't achieve a particular look cause design broke grid rules.

If no on opposes I'm going to put up a branch with this implemented and its configurable options abstracted out to a grids.css file under /variables. 
https://www.npmjs.com/package/lost#global-grid-settings

I Think we will need two tags, row and column to complement each other and our existing wrapper tag, so I'll add these as well.

**Comment by krambuhl** [comment]
I'm into having a grid system built in. I'd prefer this takes the component route over css rules directly. Building a component like `<Grid columns="4" stackAt="small" />` that can be documented and used in the same way as other components without having to learn a new css library.   

If that component uses lost under the hood, that would be a great use, but I don't like the idea of putting a whole bunch of magical new css selectors for developers to use and misuse.  I can imagine running into one of these (below) and being confused of what is actually happening, which makes debugging less clear.

```
lost-utility: edit rgb(15, 150, 25);
lost-align: right;
lost-waffle: 1/3;
lost-move: 1/2;
lost-masonry-column: 1/3 60px flex;
```

We likely don't need all of lost's functionality, so if you surface the features we _will_ use into a React component, we can keep a consistent codebase, that also makes use of Lost.  This would keep from forcing developers to understand Lost's implementation, but still be able to able to make use a good library.

**Comment by devinrayolsen** [comment]
I've just pushed my initial attempt to get us back where we at least were with out last grid system.
I agree on not needing all the features from lost and indeed going a component rout.

I've made two components, row and column.

**Row**
- <Row></Row> //Default row with max-width found in /variables/grid.css
- <Row size="reset"></Row> //full row, no max-width, no alignment
- <Row size="[1-12]"></Row> //new or us, being able to set a row's size based on grid
- <Row childPosition="top-right"></Row> //new for us, being able to position a child entitlement of row (great for banners and hero components that have textual overlay position needs)

**Column**
//Quick column classes live on
- <Column className="two-up"></Column>
- <Column className="three-up"></Column>
- <Column className="four-up"></Column>
- <Column size="[1-12]"></Column> //column count (aka 12) is configured from /variables/grid.css
- <Column gutter="false"></Column> //new for us, kills column gutters for us
- <Column push="[1-12]"></Column> //pushes element left by X number of grid columns
- <Column pull="[1-12]"></Column> //pulls element left by -X number of grid columns

With them features abstracted up for us we are back to where we were with Phoenix (far less css of course). It's also important to point out that you can still hook into both Row and Column tags via a class for a component, and apply additional lost features as seen fit but no weight added to our abstraction above.

**Comment by krambuhl** [comment]
This is a good start, I have a bunch of questions/critiques – all wildly out of order.

- How is a `<Row />` different than a `<Wrapper />`? 
- how are column's push and pull useful? this seems like edge case behavior?
- Columns className should not be related to component behaviors.  is `<Column className="two-up" />` different than `<Column size="6" />`?  (className should be reserved for consumers of the component to use)  
- The css is not using bem naming patterns.
- Using css loops is overkill, do we really use 1 columns often, let's build only parts that will be likely  used.  (100%, 50/50, 30/70 are the only grids that I see regularly used, do we want all these variations in EVERY website?)  Remove css logic where possible
- Can you explain all the different options for `<Row childPosition="*" />`? This seems like one of those things that's very tied to the Lost Implementation.  
- Take a look at the codestyle used in JSX files and try to match that style as possible.  The concatenation you are using there is very hard to read. Use destructuring for defaults instead of checking all the inputs: `push = (push > 0) ? 'push-'+push : ''`
- css variable naming patterns should be exactly the same as everything else, these variables are global, namespace them in a file with the same name: (grid.css should have variables named like: `--grid-column-breakpoint: medium;` so that when I run into a variable starting with `--grid` i know to look in the grid.css file for that information)
- It looks like you're putting in specific test styles into the styleguide, this seems like a recipe for unmaintability.  Inevitably, Someone down the line will be confused why columns and rows have borders only in the styleguide.
- It looks like your tests for Column are really testing Row.  Test columns in isolation.  (https://github.com/connectivedx/fuzzy-chainsaw/blob/feature/fz-110/source/tags/Column/Column.test.jsx) If both of these components are that intertwind, they should be merged into a single component. something like `<Grid />` and `<Grid_Row />` and `<Grid_Column />` will show that they are linked at the hip.

**Comment by devinrayolsen** [comment]
The goal with these tags is to bring our components a responsive grid approach, but do so in a way that will standardize our grid practices across our projects. Allowing the lost API to be used as needed in "grid like" component's style (or their required unique tag's style) will get unwieldy, and lead down that road of fractured grid practices across projects. 

Abstracting the basic API into these tags indeed standard the practice of the grid, but (unlike our old system) we can still use the grid API out side of these tags for corner cases (or further use of lost's API).

> How is a Row different than a Wrapper?

Other an alignment, center left and right (when size is applied) nothing. I can roll this into wrapper as it make more sense there.

> how are column's push and pull useful? this seems like edge case behavior?

Visual order vs. source order, but indeed for corner cases. We can allow developers to be big devs and use the grid API for this feature / drop it from the tag.

> Columns className should not be related to component behaviors. is <Column className="two-up" /> different than <Column size="6" />? (className should be reserved for consumers of the component to use)

Carried over from old system when we would either hide column sizes within our selectors using the grid mixins, or use a class to be used for quick columns in like CMS. We can drop what are called "quick column classes", as its quicker to write size="X" heh.

> The css is not using bem naming patterns.

updating

> Using css loops is overkill, do we really use 1 columns often, let's build only parts that will be likely used. (100%, 50/50, 30/70 are the only grids that I see regularly used, do we want all these variations in EVERY website?) Remove css logic where possible

If the goal is to provide a standard approach to our responsive grid practices, then limit size or trying to predict sizes is going to be counter productive to the overall goal. This will encourages basic grid API usage in more areas that the column tag may not be able accommodate. I'm hesitant to try and predict these or limit these sizes.

> Can you explain all the different options for <Row childPosition="*" />? This seems like one of those things that's very tied to the Lost Implementation. 

More less for overlays on banners or heros. Takes immediate child of a row, absolutely positions it within a fixed height row. This feature I'm absolutely removing, I thought it was a different kind of alignment  at first and left in for kicks.

> Take a look at the codestyle used in JSX files and try to match that style as possible. The concatenation you are using there is very hard to read. Use destructuring for defaults instead of checking all the inputs: push = (push > 0) ? 'push-'+push : ''

updating

> css variable naming patterns should be exactly the same as everything else, these variables are global, namespace them in a file with the same name: (grid.css should have variables named like: --grid-column-breakpoint: medium; so that when I run into a variable starting with --grid i know to look in the grid.css file for that information)

updating

> It looks like you're putting in specific test styles into the styleguide, this seems like a recipe for unmaintability. Inevitably, Someone down the line will be confused why columns and rows have borders only in the styleguide.

Placeholder images instead maybe? I'm trying to visually show the columns without the use of content blocks in a few areas.

> It looks like your tests for Column are really testing Row. Test columns in isolation. (https://github.com/connectivedx/fuzzy-chainsaw/blob/feature/fz-110/source/tags/Column/Column.test.jsx) If both of these components are that intertwind, they should be merged into a single component. something like <Grid /> and <Grid_Row /> and <Grid_Column /> will show that they are linked at the hip.

Right, more less wanted to show something in rows and while building them and got excited to use columns 💃 I'm going to be dropping the position tests below, and I'll just drop in a large content block instead of use columns in the row (what will be wrapper) test now.

**Comment by krambuhl** [comment]
All the rational above is sound.  Going to answer a couple below:

> If the goal is to provide a standard approach to our responsive grid practices, then limiting sizes or trying to predict sizes is going to be counter productive to the overall goal. This will encourages basic grid API usage in more areas that the column tag may not be able accommodate. I'm hesitant to try and predict these or limit these sizes as we have already been down this road when battle testing the old framework.

You are right on this one.   Let's find out what we are using in practice first.

> More less for overlays on banners or heros. Takes immediate child of a row, absolutely positions it within a fixed height row. This feature I'm absolutely removing, I thought it was a different kind of alignment at first and left in for kicks.

This seems like a useful feature for alignment.  I think some of these would be very useful, we should decide if any of them should be included by default.

> Placeholder images instead maybe? I'm trying to visually show the columns without the use of content blocks in a few areas.

Yea, this is an interesting one as the component if for layout and has no content to test with.  Maybe adding styles for these weird outliers is appropriate.  It's certainly not something we want on every component test, but on these layout helpers, maybe it's good.

**Comment by drolsen** [comment]
> You are right on this one. Let's find out what we are using in practice first.
You know..., I did pick a flavor of postcss-for that would allows us to do:

`--column--size--excludes: 1 2 3;`

So that we can customize the allowed column sizes and shave the sheet down after knowing all our used size at site completion.

> 
Yea, this is an interesting one as the component if for layout and has no content to test with. Maybe adding styles for these weird outliers is appropriate. It's certainly not something we want on every component test, but on these layout helpers, maybe it's good.

I actually ended up just putting lorem ipsum into the test as children worked fine and is more representative of real world case than visual borders ;)
